BasiliskII/src/Unix/ether_unix.cpp

```diff
@@ -240,11 +240,12 @@ bool ether_init(void)
 
 	// Determine Ethernet device type
 	net_if_type = -1;
-	if (strncmp(name, "tap", 3) == 0)
-		net_if_type = NET_IF_ETHERTAP;
 #if ENABLE_TUNTAP
-	else if (strcmp(name, "tun") == 0)
```

Comment:
What happened to "tun"?

Reply:
Sorry, this is a bit of a wall of text, hopefully the headings help a little:

### `tun` is a confusing name since it's a `tap` device

As mentioned in the blog post @hfmanson linked to, and [this more obviously canonical kernel.org source](https://www.kernel.org/doc/Documentation/networking/tuntap.txt) (which I'm quoting from):

    Depending on the type of device chosen the userspace program has to read/write
    IP packets (with tun) or ethernet frames (with tap). Which one is being used
    depends on the flags given with the ioctl().

As you can see in the code, it uses `ifr.ifr_flags = IFF_TAP`, so it's using a `tap`-type interface (thankfully, because I want to use AppleTalk over one of these interfaces!), but the code being deleted above requires you to set `ether tun` in your configuration to select this, which is confusing.

This is why @hfmanson also says "tunconfig (which should be renamed tapconfig IMO) [...]".

### Prior to this patch, `tap` is used to refer to ethertap

I guess that the term `tun` is used to refer to this type of interface because it is a "TUN/TAP" interface (as per the kernel.org document linked above) and although the code is using the `tap` type, the code also supports `ethertap` interfaces by setting `ether tapN`, so the term `tap` was already taken.

However, it's quite confusing to use `tun` to refer to these `tap` devices.  It makes it difficult to search the web for help with issues if you think you're using a `tun` device when you're using a `tap` device.

### This patch makes TUN/TAP and ethertap mutually exclusive at build time

As @hfmanson says, "when tuntap is detected in the configure script ethertap is no longer used and the user can enter tapN as ether device".  In other words, this patch means that when you run `configure`, if your host supports both ethertap and TUN/TAP, then you'll only be able to use TUN/TAP.  That means `tap` can be used to refer to the interface regardless of whether you're using ethertap or TUN/TAP.

I think that makes sense - http://user-mode-linux.sourceforge.net/old/UserModeLinux-HOWTO-6.html indicates that ethertap is for users of the Linux kernel version 2.2, and TUN/TAP is for 2.4 and later.  It also says "Ethertap is available on 2.4 and works fine. TUN/TAP is preferred to it because it has better performance and ethertap is officially considered obsolete in 2.4."  CentOS 7's kernel (version 3.10, which is relatively old) seems to be built without ethertap support.

### How to select TUN/TAP or ethertap in the future

The patch causes both ethertap and TUN/TAP to be selected by providing the name of an existing `tapN` interface.

Since the handling of TUN/TAP is changed significantly by the patch - the `tunconfig` script is called to create and configure the interface *before* BasiliskII attaches to it, instead of the previous behaviour where the `tunconfig` script was called to configure the interface *after* BasiliskII had created it, and therefore the script is significantly different - any users that previously used `tun` are going to have to do some work to move to the new scheme.

I think that the code ought to still do something about users with `tun` in their configuration, though.  With this patch, it looks to me like that configuration would cause the code to fall back to the `NET_IF_SHEEPNET` case, and presumably the user would then get an error message saying that there's no interface called `tun`.  It would be much nicer to give them an error message saying that they should read the updated `README` file.  **Also, there really needs to be an update to the `README` file!**

### Alternatives

In Bochs, you specify that you want to use a `tap` type of TUN/TAP interface by specifying `tuntap` (I guess it is implicit that it's using a `tap` type).  You also need to provide the path `/dev/net/tun` (which you can see BasiliskII hard-codes, which is probably fine).  Then, if you have an existing `tap` interface, you can append `:tapN` to that path (probably the interface's name doesn't even need to start with `tap`, as the kernel doesn't enforce that, but I haven't tested this), otherwise Bochs will create the interface for you, much like BasiliskII does without this patch.

It would probably be nice to act like Bochs in terms of separating the "I want to use a `tap`-type interface" configuration and the "here is the name of the existing `tap`-type interface I want to use" configuration rather than enforcing that the interface's name start with `tap`.

As for emulating Bochs's ability to optionally create the interface for you, the fact that this patch runs the `tunconfig` script *before* it attaches to the interface would make this tricky, since you'd want to run it *after* if you just created the interface.  I think that the scheme used in this patch, where the script is run before attaching to the interface, is pretty unconventional for emulators, and that generally if you wanted BasiliskII to not create the interface itself using `cap_net_admin`, you would simply create the interface before starting BasiliskII, which one could of course do with a wrapper script.  I would prefer to have something that worked more like Bochs and less like this patch in this regard, although I don't have strong feelings either way - I'm happy to just create the interface myself and then tell BasiliskII to run an empty script.